File: c/LeetCode_257_BinaryTreePaths.c
#include <stdio.h>
#include <stdlib.h>


struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


char** binaryTreePaths(struct TreeNode* root, int* returnSize) {
    if (root == NULL) return NULL;

    int *paths = (int*) malloc(10001 * sizeof(int*));
    struct TreeNode** stack = (struct TreeNode**) malloc(10001 * sizeof(struct TreeNode*));
    int inx = 0;
    struct TreeNode* curr = root; 
    struct TreeNode* prev = NULL;

    while (1) {
        while (curr) {
            stack[inx] = curr; 
            paths[inx] = curr->val;
            curr = curr->left;
            ++inx;
        }

        curr = stack[--inx];

        if (curr->right && curr->right != prev) {
            curr = curr->right;
        } else {
            if (!curr->right && !curr->left) {
                printf("%d ", paths[0]);
            } 
        }
    }

}

File: python/LeetCode_257_BinaryTreePaths.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def binaryTreePaths(root):
    if (root == None): return [""]
    
    path = []
    stack = []
    curr = root
    prev = None
 
    result = []
    while curr or stack:
        while curr:
            stack.append(curr)
            path.append(curr.val)
            curr = curr.left
 
        curr = stack[-1]
 
        if curr.right and curr.right != prev:
            curr = curr.right
        else:
            if not curr.right and not curr.left:
                result.append( "->".join(map(str, path)))
 
            stack.pop()
            path.pop()
            prev = curr
            curr = None
    
    return result

